File: python/2018/day1.py
# vi: set shiftwidth=4 tabstop=4 expandtab:
import datetime
import os

top_dir = os.path.dirname(os.path.abspath(__file__)) + "/../../"


def get_freq_from_file(file_path=top_dir + "resources/year2018_day1_input.txt"):
    with open(file_path) as f:
        return [int(l.strip()) for l in f]


def get_resulting_freq(freqs):
    return sum(freqs)


def get_first_repetition(freqs):
    freq = 0
    seens = set([freq])
    while True:
        for f in freqs:
            freq += f
            if freq in seens:
                return freq
            seens.add(freq)


def run_tests():
    assert get_resulting_freq([+1, -2, +3, +1]) == 3
    assert get_first_repetition([+1, -2, +3, +1]) == 2


def get_solutions():
    freqs = get_freq_from_file()
    print(get_resulting_freq(freqs) == 587)
    print(get_first_repetition(freqs) == 83130)


if __name__ == "__main__":
    begin = datetime.datetime.now()
    run_tests()
    get_solutions()
    end = datetime.datetime.now()
    print(end - begin)

File: rust/src/2018/day1.rs
// vi: set shiftwidth=4 tabstop=4 expandtab:
use common::input::check_answer;
use common::input::collect_from_lines;
use common::input::get_answers;
use common::input::get_file_content;
use std::collections::HashSet;
use std::time::Instant;

const INPUT_FILEPATH: &str = "../resources/year2018_day1_input.txt";
const ANSWERS_FILEPATH: &str = "../resources/year2018_day1_answer.txt";

type Int = i32;
type InputContent = Vec<Int>;

fn get_input_from_str(string: &str) -> InputContent {
    collect_from_lines(string)
}

fn part1(arg: &InputContent) -> Int {
    arg.iter().sum()
}

fn part2(arg: &InputContent) -> Int {
    let mut f: Int = 0;
    let mut frequencies = HashSet::new();
    loop {
        for d in arg {
            f += d;
            if frequencies.contains(&f) {
                return f;
            }
            frequencies.insert(f);
        }
    }
}

fn main() {
    let before = Instant::now();
    let data = get_input_from_str(&get_file_content(INPUT_FILEPATH));
    let (ans, ans2) = get_answers(ANSWERS_FILEPATH);
    let solved = true;
    let res = part1(&data);
    check_answer(&res.to_string(), ans, solved);
    let res2 = part2(&data);
    check_answer(&res2.to_string(), ans2, solved);
    println!("Elapsed time: {:.2?}", before.elapsed());
}

#[cfg(test)]
mod tests {
    use super::*;

    const EXAMPLE: &str = "+1
-2
+3
+1";

    #[test]
    fn test_part1() {
        assert_eq!(part1(&get_input_from_str(EXAMPLE)), 3);
    }

    #[test]
    fn test_part2() {
        assert_eq!(part2(&get_input_from_str(EXAMPLE)), 2);
    }
}
